// zigbee-herdsman/dist/adapter/z-stack/znp/parameterType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ParameterType;
(function (ParameterType) {
    ParameterType[ParameterType["UINT8"] = 0] = "UINT8";
    ParameterType[ParameterType["UINT16"] = 1] = "UINT16";
    ParameterType[ParameterType["UINT32"] = 2] = "UINT32";
    ParameterType[ParameterType["IEEEADDR"] = 3] = "IEEEADDR";
    ParameterType[ParameterType["BUFFER"] = 4] = "BUFFER";
    ParameterType[ParameterType["BUFFER8"] = 5] = "BUFFER8";
    ParameterType[ParameterType["BUFFER16"] = 6] = "BUFFER16";
    ParameterType[ParameterType["BUFFER18"] = 7] = "BUFFER18";
    ParameterType[ParameterType["BUFFER32"] = 8] = "BUFFER32";
    ParameterType[ParameterType["BUFFER42"] = 9] = "BUFFER42";
    ParameterType[ParameterType["BUFFER100"] = 10] = "BUFFER100";
    ParameterType[ParameterType["LIST_UINT8"] = 11] = "LIST_UINT8";
    ParameterType[ParameterType["LIST_UINT16"] = 12] = "LIST_UINT16";
    ParameterType[ParameterType["LIST_ROUTING_TABLE"] = 13] = "LIST_ROUTING_TABLE";
    ParameterType[ParameterType["LIST_BIND_TABLE"] = 14] = "LIST_BIND_TABLE";
    ParameterType[ParameterType["LIST_NEIGHBOR_LQI"] = 15] = "LIST_NEIGHBOR_LQI";
    ParameterType[ParameterType["LIST_NETWORK"] = 16] = "LIST_NETWORK";
    ParameterType[ParameterType["LIST_ASSOC_DEV"] = 17] = "LIST_ASSOC_DEV";
    ParameterType[ParameterType["INT8"] = 18] = "INT8";
})(ParameterType || (ParameterType = {}));
exports.default = ParameterType;
