// zigbee-herdsman/dist/adapter/z-stack/adapter/backup.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Restore = exports.Backup = void 0;
const Constants = __importStar(require("../constants"));
const tstype_1 = require("./tstype");
const constants_1 = require("../unpi/constants");
const fs_1 = __importDefault(require("fs"));
const es6_1 = __importDefault(require("fast-deep-equal/es6"));
const nvItems_1 = __importDefault(require("./nvItems"));
const { NvItemsIds, NvSystemIds, ZnpCommandStatus } = Constants.COMMON;
const PRODUCT_ALL = -1;
const items = {
    ZCD_NV_EXTADDR: {
        id: NvItemsIds.EXTADDR,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_NIB: {
        id: NvItemsIds.NIB,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_PANID: {
        id: NvItemsIds.PANID,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_EXTENDED_PAN_ID: {
        id: NvItemsIds.EXTENDED_PAN_ID,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_NWK_ACTIVE_KEY_INFO: {
        id: NvItemsIds.NWK_ACTIVE_KEY_INFO,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_NWK_ALTERN_KEY_INFO: {
        id: NvItemsIds.NWK_ALTERN_KEY_INFO,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_APS_USE_EXT_PANID: {
        id: NvItemsIds.APS_USE_EXT_PANID,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_PRECFGKEY: {
        id: NvItemsIds.PRECFGKEY,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_PRECFGKEY_ENABLE: {
        id: NvItemsIds.PRECFGKEYS_ENABLE,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_CHANLIST: {
        id: NvItemsIds.CHANLIST,
        offset: 0x00,
        osal: true,
        product: PRODUCT_ALL,
    },
    ZCD_NV_LEGACY_TCLK_TABLE_START: {
        id: NvItemsIds.LEGACY_TCLK_TABLE_START,
        product: tstype_1.ZnpVersion.zStack30x,
        offset: 0x00,
        osal: true,
    },
    ZCD_NV_EX_TCLK_TABLE: {
        sysid: NvSystemIds.ZSTACK,
        itemid: NvItemsIds.EX_TCLK_TABLE,
        subid: 0,
        product: tstype_1.ZnpVersion.zStack3x0,
        osal: false,
        offset: 0x00
    },
    ZCD_NV_LEGACY_NWK_SEC_MATERIAL_TABLE_START: {
        id: NvItemsIds.LEGACY_NWK_SEC_MATERIAL_TABLE_START,
        product: tstype_1.ZnpVersion.zStack30x,
        offset: 0x00,
        osal: true,
    },
    ZCD_NV_EX_NWK_SEC_MATERIAL_TABLE: {
        sysid: NvSystemIds.ZSTACK,
        itemid: NvItemsIds.EX_NWK_SEC_MATERIAL_TABLE,
        subid: 0,
        product: tstype_1.ZnpVersion.zStack3x0,
        osal: false,
        offset: 0x00,
    }
};
function Backup(znp) {
    return __awaiter(this, void 0, void 0, function* () {
        const product = (yield znp.request(constants_1.Subsystem.SYS, 'version', {})).payload.product;
        if (product !== tstype_1.ZnpVersion.zStack30x && product !== tstype_1.ZnpVersion.zStack3x0) {
            throw new Error('Backup is only supported for Z-Stack 3');
        }
        // eslint-disable-next-line
        const data = {};
        for (const [key, entry] of Object.entries(items)) {
            if (entry.product === PRODUCT_ALL || entry.product === product) {
                let result = null;
                if (entry.osal) {
                    result = yield znp.request(constants_1.Subsystem.SYS, 'osalNvRead', entry);
                }
                else {
                    const len = (yield znp.request(constants_1.Subsystem.SYS, 'nvLength', entry)).payload.len;
                    result = yield znp.request(constants_1.Subsystem.SYS, 'nvRead', Object.assign(Object.assign({}, entry), { len }));
                }
                data[key] = Object.assign(Object.assign({}, entry), { value: [...result.payload.value], len: result.payload.value.length });
            }
        }
        return {
            adapterType: 'zStack',
            time: new Date().toUTCString(),
            meta: { product },
            data,
        };
    });
}
exports.Backup = Backup;
function Restore(znp, backupPath, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const backup = JSON.parse(fs_1.default.readFileSync(backupPath).toString());
        const product = (yield znp.request(constants_1.Subsystem.SYS, 'version', {})).payload.product;
        if (backup.adapterType !== 'zStack') {
            throw new Error(`Cannot restore backup, backup is for '${backup.adapterType}', current is 'zStack'`);
        }
        if (backup.meta.product != product) {
            throw new Error(`Cannot restore backup, backup is for '${tstype_1.ZnpVersion[backup.meta.product]}', ` +
                `current is '${tstype_1.ZnpVersion[product]}'`);
        }
        if (!es6_1.default(backup.data.ZCD_NV_CHANLIST.value, Constants.Utils.getChannelMask(options.channelList))) {
            throw new Error(`Cannot restore backup, channel of backup is different`);
        }
        if (!es6_1.default(backup.data.ZCD_NV_PRECFGKEY.value, options.networkKey)) {
            throw new Error(`Cannot restore backup, networkKey of backup is different`);
        }
        if (!es6_1.default(backup.data.ZCD_NV_PANID.value, Array.from(nvItems_1.default.panID(options.panID).value))) {
            throw new Error(`Cannot restore backup, panID of backup is different`);
        }
        if (!es6_1.default(backup.data.ZCD_NV_EXTENDED_PAN_ID.value, options.extendedPanID)) {
            throw new Error(`Cannot restore backup, extendedPanID of backup is different`);
        }
        const ZCD_NV_NIB = Object.assign(Object.assign({}, backup.data.ZCD_NV_NIB), { initvalue: backup.data.ZCD_NV_NIB.value, initlen: backup.data.ZCD_NV_NIB.len });
        const bdbNodeIsOnANetwork = {
            id: NvItemsIds.BDBNODEISONANETWORK,
            len: 0x01,
            offset: 0x0,
            value: [0x01],
            initlen: 0x01,
            initvalue: [0x01]
        };
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_EXTADDR);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvItemInit', ZCD_NV_NIB, null, [ZnpCommandStatus.NV_ITEM_INITIALIZED]);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_PANID);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_EXTENDED_PAN_ID);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_NWK_ACTIVE_KEY_INFO);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_NWK_ALTERN_KEY_INFO);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_APS_USE_EXT_PANID);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_PRECFGKEY);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_PRECFGKEY_ENABLE);
        if (product === tstype_1.ZnpVersion.zStack30x) {
            yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_LEGACY_TCLK_TABLE_START);
            yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', backup.data.ZCD_NV_LEGACY_NWK_SEC_MATERIAL_TABLE_START);
        }
        else {
            /* istanbul ignore else */
            if (product === tstype_1.ZnpVersion.zStack3x0) {
                yield znp.request(constants_1.Subsystem.SYS, 'nvWrite', backup.data.ZCD_NV_EX_TCLK_TABLE);
                yield znp.request(constants_1.Subsystem.SYS, 'nvWrite', backup.data.ZCD_NV_EX_NWK_SEC_MATERIAL_TABLE);
            }
        }
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvItemInit', nvItems_1.default.znpHasConfiguredInit(product), null, [ZnpCommandStatus.NV_ITEM_INITIALIZED]);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', nvItems_1.default.znpHasConfigured(product));
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvItemInit', bdbNodeIsOnANetwork, null, [ZnpCommandStatus.SUCCESS, ZnpCommandStatus.NV_ITEM_INITIALIZED]);
        yield znp.request(constants_1.Subsystem.SYS, 'osalNvWrite', bdbNodeIsOnANetwork);
        yield znp.request(constants_1.Subsystem.SYS, 'resetReq', { type: Constants.SYS.resetType.SOFT });
    });
}
exports.Restore = Restore;
