// typedoc/dist/lib/models/types/type-parameter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeParameterType = void 0;
const abstract_1 = require("./abstract");
class TypeParameterType extends abstract_1.Type {
    constructor(name) {
        super();
        this.type = 'typeParameter';
        this.name = name;
    }
    clone() {
        const clone = new TypeParameterType(this.name);
        clone.constraint = this.constraint;
        clone.default = this.default;
        return clone;
    }
    equals(type) {
        if (!(type instanceof TypeParameterType)) {
            return false;
        }
        let constraintEquals = false;
        if (this.constraint && type.constraint) {
            constraintEquals = type.constraint.equals(this.constraint);
        }
        else if (!this.constraint && !type.constraint) {
            constraintEquals = true;
        }
        let defaultEquals = false;
        if (this.default && type.default) {
            defaultEquals = type.default.equals(this.default);
        }
        else if (!this.default && !type.default) {
            defaultEquals = true;
        }
        return constraintEquals && defaultEquals;
    }
    toString() {
        return this.name;
    }
}
exports.TypeParameterType = TypeParameterType;
