// zigbee-herdsman/dist/zcl/definition/dataType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DataType;
(function (DataType) {
    // TODO: refactor to snake casing
    DataType[DataType["noData"] = 0] = "noData";
    DataType[DataType["data8"] = 8] = "data8";
    DataType[DataType["data16"] = 9] = "data16";
    DataType[DataType["data24"] = 10] = "data24";
    DataType[DataType["data32"] = 11] = "data32";
    DataType[DataType["data40"] = 12] = "data40";
    DataType[DataType["data48"] = 13] = "data48";
    DataType[DataType["data56"] = 14] = "data56";
    DataType[DataType["data64"] = 15] = "data64";
    DataType[DataType["boolean"] = 16] = "boolean";
    DataType[DataType["bitmap8"] = 24] = "bitmap8";
    DataType[DataType["bitmap16"] = 25] = "bitmap16";
    DataType[DataType["bitmap24"] = 26] = "bitmap24";
    DataType[DataType["bitmap32"] = 27] = "bitmap32";
    DataType[DataType["bitmap40"] = 28] = "bitmap40";
    DataType[DataType["bitmap48"] = 29] = "bitmap48";
    DataType[DataType["bitmap56"] = 30] = "bitmap56";
    DataType[DataType["bitmap64"] = 31] = "bitmap64";
    DataType[DataType["uint8"] = 32] = "uint8";
    DataType[DataType["uint16"] = 33] = "uint16";
    DataType[DataType["uint24"] = 34] = "uint24";
    DataType[DataType["uint32"] = 35] = "uint32";
    DataType[DataType["uint40"] = 36] = "uint40";
    DataType[DataType["uint48"] = 37] = "uint48";
    DataType[DataType["uint56"] = 38] = "uint56";
    DataType[DataType["uint64"] = 39] = "uint64";
    DataType[DataType["int8"] = 40] = "int8";
    DataType[DataType["int16"] = 41] = "int16";
    DataType[DataType["int24"] = 42] = "int24";
    DataType[DataType["int32"] = 43] = "int32";
    DataType[DataType["enum8"] = 48] = "enum8";
    DataType[DataType["enum16"] = 49] = "enum16";
    DataType[DataType["singlePrec"] = 57] = "singlePrec";
    DataType[DataType["doublePrec"] = 58] = "doublePrec";
    DataType[DataType["octetStr"] = 65] = "octetStr";
    DataType[DataType["charStr"] = 66] = "charStr";
    DataType[DataType["longOctetStr"] = 67] = "longOctetStr";
    DataType[DataType["longCharStr"] = 68] = "longCharStr";
    DataType[DataType["array"] = 72] = "array";
    DataType[DataType["struct"] = 76] = "struct";
    DataType[DataType["set"] = 80] = "set";
    DataType[DataType["bag"] = 81] = "bag";
    DataType[DataType["tod"] = 224] = "tod";
    DataType[DataType["date"] = 225] = "date";
    DataType[DataType["utc"] = 226] = "utc";
    DataType[DataType["clusterId"] = 232] = "clusterId";
    DataType[DataType["attrId"] = 233] = "attrId";
    DataType[DataType["bacOid"] = 234] = "bacOid";
    DataType[DataType["ieeeAddr"] = 240] = "ieeeAddr";
    DataType[DataType["secKey"] = 241] = "secKey";
    DataType[DataType["unknown"] = 255] = "unknown";
    // TODO: ones below are not in BuffaloZcl and thus cannot be parsed/written
    DataType[DataType["int40"] = 44] = "int40";
    DataType[DataType["int48"] = 45] = "int48";
    DataType[DataType["int56"] = 46] = "int56";
    DataType[DataType["int64"] = 47] = "int64";
    DataType[DataType["semiPrec"] = 56] = "semiPrec";
})(DataType || (DataType = {}));
exports.default = DataType;
