// zigbee-herdsman/dist/zcl/definition/status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Status;
(function (Status) {
    Status[Status["SUCCESS"] = 0] = "SUCCESS";
    Status[Status["FAILURE"] = 1] = "FAILURE";
    Status[Status["NOT_AUTHORIZED"] = 126] = "NOT_AUTHORIZED";
    Status[Status["RESERVED_FIELD_NOT_ZERO"] = 127] = "RESERVED_FIELD_NOT_ZERO";
    Status[Status["MALFORMED_COMMAND"] = 128] = "MALFORMED_COMMAND";
    Status[Status["UNSUP_CLUSTER_COMMAND"] = 129] = "UNSUP_CLUSTER_COMMAND";
    Status[Status["UNSUP_GENERAL_COMMAND"] = 130] = "UNSUP_GENERAL_COMMAND";
    Status[Status["UNSUP_MANUF_CLUSTER_COMMAND"] = 131] = "UNSUP_MANUF_CLUSTER_COMMAND";
    Status[Status["UNSUP_MANUF_GENERAL_COMMAND"] = 132] = "UNSUP_MANUF_GENERAL_COMMAND";
    Status[Status["INVALID_FIELD"] = 133] = "INVALID_FIELD";
    Status[Status["UNSUPPORTED_ATTRIBUTE"] = 134] = "UNSUPPORTED_ATTRIBUTE";
    Status[Status["INVALID_VALUE"] = 135] = "INVALID_VALUE";
    Status[Status["READ_ONLY"] = 136] = "READ_ONLY";
    Status[Status["INSUFFICIENT_SPACE"] = 137] = "INSUFFICIENT_SPACE";
    Status[Status["DUPLICATE_EXISTS"] = 138] = "DUPLICATE_EXISTS";
    Status[Status["NOT_FOUND"] = 139] = "NOT_FOUND";
    Status[Status["UNREPORTABLE_ATTRIBUTE"] = 140] = "UNREPORTABLE_ATTRIBUTE";
    Status[Status["INVALID_DATA_TYPE"] = 141] = "INVALID_DATA_TYPE";
    Status[Status["INVALID_SELECTOR"] = 142] = "INVALID_SELECTOR";
    Status[Status["WRITE_ONLY"] = 143] = "WRITE_ONLY";
    Status[Status["INCONSISTENT_STARTUP_STATE"] = 144] = "INCONSISTENT_STARTUP_STATE";
    Status[Status["DEFINED_OUT_OF_BAND"] = 145] = "DEFINED_OUT_OF_BAND";
    Status[Status["INCONSISTENT"] = 146] = "INCONSISTENT";
    Status[Status["ACTION_DENIED"] = 147] = "ACTION_DENIED";
    Status[Status["TIMEOUT"] = 148] = "TIMEOUT";
    Status[Status["ABORT"] = 149] = "ABORT";
    Status[Status["INVALID_IMAGE"] = 150] = "INVALID_IMAGE";
    Status[Status["WAIT_FOR_DATA"] = 151] = "WAIT_FOR_DATA";
    Status[Status["NO_IMAGE_AVAILABLE"] = 152] = "NO_IMAGE_AVAILABLE";
    Status[Status["REQUIRE_MORE_IMAGE"] = 153] = "REQUIRE_MORE_IMAGE";
    Status[Status["NOTIFICATION_PENDING"] = 154] = "NOTIFICATION_PENDING";
    Status[Status["HARDWARE_FAILURE"] = 192] = "HARDWARE_FAILURE";
    Status[Status["SOFTWARE_FAILURE"] = 193] = "SOFTWARE_FAILURE";
    Status[Status["CALIBRATION_ERROR"] = 194] = "CALIBRATION_ERROR";
    Status[Status["UNSUPPORTED_CLUSTER"] = 195] = "UNSUPPORTED_CLUSTER";
})(Status || (Status = {}));
exports.default = Status;
