// typedoc/dist/lib/converter/factories/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTypeParameter = exports.createSignature = exports.createReferenceType = exports.createParameter = exports.createDeclaration = exports.createComment = void 0;
var comment_1 = require("./comment");
Object.defineProperty(exports, "createComment", { enumerable: true, get: function () { return comment_1.createComment; } });
var declaration_1 = require("./declaration");
Object.defineProperty(exports, "createDeclaration", { enumerable: true, get: function () { return declaration_1.createDeclaration; } });
var parameter_1 = require("./parameter");
Object.defineProperty(exports, "createParameter", { enumerable: true, get: function () { return parameter_1.createParameter; } });
var reference_1 = require("./reference");
Object.defineProperty(exports, "createReferenceType", { enumerable: true, get: function () { return reference_1.createReferenceType; } });
var signature_1 = require("./signature");
Object.defineProperty(exports, "createSignature", { enumerable: true, get: function () { return signature_1.createSignature; } });
var type_parameter_1 = require("./type-parameter");
Object.defineProperty(exports, "createTypeParameter", { enumerable: true, get: function () { return type_parameter_1.createTypeParameter; } });
